librz/analysis/arch/arm/arm_il32.c

```diff
@@ -2314,22 +2515,880 @@ static RzILOpEffect *tbb(cs_insn *insn, bool is_thumb) {
+			RzILOpBitVector *rt1_val = REG(1);
+			RzILOpBitVector *rt2_val = REG(2);
+			if (!rt1_val || !rt2_val) {
+				return NULL;
+			}
+			return write_reg(REGID(0), APPEND(rt2_val, rt1_val));
+		}
+		// vmov <Rt1> <Rt2> <Dm>, Rt1 = Dm[low], Rt2 = Dm[high]
+		RzILOpBitVector *reg_val = REG(2);
+		if (!reg_val) {
+			return NULL;
+		}
+		RzILOpBitVector *rt1_val = UNSIGNED(32, reg_val);
+		RzILOpBitVector *rt2_val = UNSIGNED(32, SHIFTR0(UN(8, 32), reg_val));
+		return SEQ2(write_reg(REGID(0), rt1_val),
+			write_reg(REGID(1), rt2_val));
+	}
+
+	// 2 core registers and 2 single-word registers
+	// vmov <Sm1> <Sm2> <Rt1> <Rt2>
+	// vmov <Rt1> <Rt2> <Sm1> <Sm2>
+	if (OPCOUNT() == 4) {
+		RzILOpBitVector *rt1_val = REG(2);
+		RzILOpBitVector *rt2_val = REG(3);
+		if (!rt1_val || !rt2_val) {
+			return NULL;
+		}
+		return SEQ2(write_reg(REGID(0), rt1_val),
+			write_reg(REGID(1), rt2_val));
+	}
+
+	// core register to scalar
+	if (NEON_LANE(0) != -1 && !is_core_reg(REGID(0)) && is_core_reg(REGID(1))) {
+		// vmov.<vec_size> <Dd>[lane], <Rt>
+		// <Dd>[lane] = <Rt>{vecsize - 1 : 0}
+		if (!VVEC_SIZE(insn) || NEON_LANE(0) == -1) {
+			return NULL;
+		}
+		RzILOpBitVector *rt_val = UNSIGNED(VVEC_SIZE(insn), REG(1));
+		return write_reg_lane(REGID(0), NEON_LANE(0), VVEC_SIZE(insn), rt_val);
+	}
+
+	// scalar to core register
+	if (NEON_LANE(1) != -1 && !is_core_reg(REGID(1)) && is_core_reg(REGID(0))) {
+		// vmov.<dt> <Rt> <Dd>[lane]
+		// <Rt> = extend_to_32(<Dd>[lane], lane has size of dt)
+		// unsigned/signed extend is specified by <dt> in capstone
+		if (VVEC_DT(insn) == ARM_VECTORDATA_INVALID) {
+			return NULL;
+		}
+		bool use_zero_ext = true;
+		if (VVEC_DT(insn) == ARM_VECTORDATA_S8 || VVEC_DT(insn) == ARM_VECTORDATA_S16) {
+			use_zero_ext = false;
+		}
+		RzILOpBitVector *lane_val = read_reg_lane(REGID(1), NEON_LANE(1), DT_WIDTH(insn));
+		RzILOpBitVector *ext_lane_val = use_zero_ext ? UNSIGNED(32, lane_val) : SIGNED(32, lane_val);
+		return write_reg(REGID(0), ext_lane_val);
+	}
+
+	// 1. vmov rd, rn
+	// 2. core register and single-word register
 	RzILOpBitVector *val = ARG(1);
 	if (!val) {
 		return NULL;
 	}
+
+	// vmvn Qd, Qn
+	if (insn->id == ARM_INS_VMVN) {
+		val = LOGNOT(val);
+	}
+
 	return write_reg(REGID(0), val);
 }
 
+/**
+ * Capstone: ARM_INS_VMRS
+ * ARM: vmrs
+ * read extension/NEON system register into core register
+ */
+static RzILOpEffect *vmrs(cs_insn *insn, bool is_thumb) {
+	if (!ISREG(0) || !ISREG(1)) {
+		return NULL;
+	}
+
+	// VMRS <Rt>, FPSCR only
+	if (REGID(1) != ARM_REG_FPSCR) {
+		return NULL;
+	}
+
+	// if <Rt> is APSR, transfer to flags
+	if (REGID(1) == ARM_REG_APSR_NZCV) {
+		RzILOpBitVector *val = VARG("fpscr");
+		return SEQ4(
+			SETG("nf", INV(IS_ZERO(LOGAND(val, U32(1ul << 31))))),
+			SETG("zf", INV(IS_ZERO(LOGAND(DUP(val), U32(1ul << 30))))),
+			SETG("cf", INV(IS_ZERO(LOGAND(DUP(val), U32(1ul << 29))))),
+			SETG("vf", INV(IS_ZERO(LOGAND(DUP(val), U32(1ul << 28))))));
```

Comment:
It might be worth to move this into its own `set_flags` function. Because it repeats below.

Reply:
not really, the below one is read value from those flags. A similar function is "update_flags_from_cpsr", but we only focus on 4 flag bits instead the whole `cpsr`. So I think it's ok to keep it.